x/pairing/keeper/pairing.go

```diff
@@ -110,43 +109,74 @@ func (k Keeper) GetPairingForClient(ctx sdk.Context, chainID string, clientAddre
 	var strictestPolicy planstypes.Policy
 
-	epoch, err := k.VerifyPairingData(ctx, chainID, clientAddress, block)
+	epoch, providersType, err := k.VerifyPairingData(ctx, chainID, clientAddress, block)
 	if err != nil {
 		return nil, 0, "", fmt.Errorf("invalid pairing data: %s", err)
 	}
 
+	stakeEntries, found, epochHash := k.epochStorageKeeper.GetEpochStakeEntries(ctx, epoch, chainID)
+	if !found {
+		return nil, 0, "", fmt.Errorf("did not find providers for pairing: epoch:%d, chainID: %s", block, chainID)
+	}
+
 	project, err := k.GetProjectData(ctx, clientAddress, chainID, block)
 	if err != nil {
 		return nil, 0, "", err
 	}
 
-	strictestPolicy, allowedCU, err = k.getProjectStrictestPolicy(ctx, project, chainID)
+	strictestPolicy, allowedCU, minStake, err := k.getProjectStrictestPolicy(ctx, project, chainID)
 	if err != nil {
 		return nil, 0, "", fmt.Errorf("invalid user for pairing: %s", err.Error())
 	}
 
-	possibleProviders, found, epochHash := k.epochStorageKeeper.GetEpochStakeEntries(ctx, epoch, chainID)
-	if !found {
-		return nil, 0, "", fmt.Errorf("did not find providers for pairing: epoch:%d, chainID: %s", block, chainID)
+	if providersType == spectypes.Spec_static {
+		return stakeEntries, allowedCU, project.Index, nil
 	}
 
 	filters := pairingfilters.GetAllFilters()
 
-	possibleProviders, err = pairingfilters.FilterProviders(ctx, filters, possibleProviders, strictestPolicy, epoch)
+	stakeEntries, err = pairingfilters.FilterProviders(ctx, filters, stakeEntries, strictestPolicy, epoch)
 	if err != nil {
 		return nil, 0, "", err
 	}
 
-	providers, err = k.calculatePairingForClient(ctx, possibleProviders, project.Index, block,
-		chainID, epochHash, strictestPolicy.MaxProvidersToPair)
+	// create the pairing slots with assigned reqs
+	slots := pairingscores.CalcSlots(strictestPolicy, minStake)
+
+	// group identical slots (in terms of reqs types)
+	slotGroups := pairingscores.GroupSlots(slots)
+
+	// create providerScore array with all possible providers
+	providerScores := []*pairingscores.PairingScore{}
+	for i := range stakeEntries {
+		providerScore := pairingscores.NewPairingScore(&stakeEntries[i])
+		providerScores = append(providerScores, providerScore)
+	}
+
+	// calculate score (always on the diff in score components of consecutive groups) and pick providers
+	prevGroupSlot := pairingscores.NewPairingSlot() // init dummy slot to compare to
+	prevGroupSlot.Reqs = map[string]pairingscores.ScoreReq{}
+	indexToSkip := make(map[int]bool) // keep the indices of chosen providers to we won't pick the same providers twice (for different groups)
+	for _, group := range slotGroups {
+		diffSlot := group.Subtract(prevGroupSlot)
+		err := pairingscores.CalcPairingScore(providerScores, pairingscores.GetStrategy(), diffSlot, minStake)
+		if err != nil {
+			return nil, 0, "", err
+		}
+		hashData := pairingscores.PrepareHashData(project.Index, chainID, epochHash)
```

Comment:
why is this calculated everytime, it is the same for each iteration

Reply:
done. see https://github.com/lavanet/lava/commit/cd637d649bba13af1a7a698952aa92539cd49074